### PR TITLE
[enhancement](fe-memory) support label num threshold to reduce fe memory consumption  

### PR DESCRIPTION
**Pull request by Hastyshell**
## Proposed changes

Support label num threshold for bulk load, routine load and sync.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.15 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162069355 Bytes

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run p0

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by dataroaring** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadManager.java [hunk 1]
```diff
@@ -679,19 +681,35 @@ public void processTimeoutTasks() {
...
+     * Remove finished routine load jobs from idToRoutineLoadJob
+     * This function is called periodically if Config.label_num_threshold is set.
+     * Cancelled and stopped job will be removed.
+     */
+    public void cleanOverLimitRoutineLoadJobs() {
+        if (idToRoutineLoadJob.size() <= Config.label_num_threshold) {
+            return;
+        }
+        LOG.debug("begin to clean routine load jobs");
+        clearRoutineLoadJobIf(RoutineLoadJob::isFinal, Config.label_num_threshold);
```
Comment (yujun777) [inline fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadManager.java:700]: Is it better to remove oldest jobs when exceed limit

File: fe/fe-core/src/main/java/org/apache/doris/load/sync/SyncJobManager.java [hunk 2]
```diff
@@ -283,13 +286,28 @@ public void updateNeedSchedule() throws UserException {
...
+
+    /**
+     * Remove completed jobs if total job num exceed Config.label_num_threshold
+     */
+    public void cleanOverLimitSyncJobs() {
+        if (idToSyncJob.size() <= Config.label_num_threshold) {
+            return;
+        }
+        LOG.debug("begin to clean finished sync jobs ");
+        cleanFinishedSyncJobsIf(SyncJob::isCompleted, Config.label_num_threshold);
```
Comment (yujun777) [inline fe/fe-core/src/main/java/org/apache/doris/load/sync/SyncJobManager.java:300]: Is it better to remove oldest jobs when exceed limit

**Comment by TangSiyang2001** [comment]
I'd like to introduce a better policy for num based restriction. Eviction is better to happend after insertion, rather than in a fix interval.

File: fe/fe-core/src/main/java/org/apache/doris/load/loadv2/LoadManager.java [hunk 3]
```diff
@@ -90,7 +92,7 @@
 public class LoadManager implements Writable {
     private static final Logger LOG = LogManager.getLogger(LoadManager.class);
 
-    private Map<Long, LoadJob> idToLoadJob = Maps.newConcurrentMap();
+    private Map<Long, LoadJob> idToLoadJob = Collections.synchronizedMap(Maps.newLinkedHashMap());
     private Map<Long, Map<String, List<LoadJob>>> dbIdToLabelToLoadJobs = Maps.newConcurrentMap();
```
Comment (SWJTU-ZhangLei) [inline fe/fe-core/src/main/java/org/apache/doris/load/loadv2/LoadManager.java:96]: Why replace concurrentMap with synchronizedMap ? what is the advantages and disadvantages?

File: fe/fe-core/src/main/java/org/apache/doris/load/loadv2/LoadManager.java [hunk 4]
```diff
@@ -399,13 +401,26 @@ public long getLoadJobNum(JobState jobState, EtlJobType jobType) {
...
+        }
+        removeLoadJobIf(LoadJob::isCompleted, Config.label_num_threshold);
+    }
 
+    private void removeLoadJobIf(Predicate<LoadJob> pred, int numLimit) {
         writeLock();
         try {
             Iterator<Map.Entry<Long, LoadJob>> iter = idToLoadJob.entrySet().iterator();
-            while (iter.hasNext()) {
+            while (iter.hasNext() && idToLoadJob.size() > numLimit) {
```
Comment (SWJTU-ZhangLei) [inline fe/fe-core/src/main/java/org/apache/doris/load/loadv2/LoadManager.java:421]: do we need think about sorting by finishTime ? I think we keep the last recently label_num_threshold final job maybe better

File: fe/fe-core/src/test/java/org/apache/doris/load/routineload/RoutineLoadManagerTest.java [hunk 5]
```diff
@@ -801,7 +801,7 @@ public void testCleanOldRoutineLoadJobs(@Injectable RoutineLoadJob routineLoadJo
 
         new Expectations() {
             {
-                routineLoadJob.needRemove();
+                routineLoadJob.isExpired();
```
Comment (SWJTU-ZhangLei) [inline fe/fe-core/src/test/java/org/apache/doris/load/routineload/RoutineLoadManagerTest.java:804]: lack of lots of jobs ut test and regression test

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
ut is ongoing

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Review by yujun777 — APPROVED** [review]
LGTM

**Comment by dataroaring** [comment]
run buildall

**Comment by SWJTU-ZhangLei** [comment]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**